components/merkledb/src/db.rs

```diff
@@ -209,51 +228,102 @@ impl WorkingPatch {
         self.changes.borrow().is_empty()
     }
 
-    fn take_view_changes(&self, address: &IndexAddress) -> Option<ViewChanges> {
+    /// Takes a cell with changes for a specific `View` out of the patch.
+    /// The returned cell is guaranteed to contain an `Rc` with an exclusive ownership.
+    fn take_view_changes(&self, address: &IndexAddress) -> ChangesCell {
         let view_changes = {
             let mut changes = self.changes.borrow_mut();
             let view_changes = changes.get_mut(address).map(Option::take);
             view_changes.unwrap_or_else(|| {
                 changes
                     .entry(address.clone())
-                    .or_insert_with(|| Some(ViewChanges::new()))
+                    .or_insert_with(|| Some(Rc::new(ViewChanges::new())))
                     .take()
             })
         };
 
-        assert!(
-            view_changes.is_some(),
-            "multiple mutable borrows of an index at {:?}",
-            address
-        );
+        if let Some(ref view_changes) = view_changes {
+            assert!(
+                Rc::strong_count(view_changes) == 1,
+                "Attempting to borrow {:?} mutably while it's borrowed immutably",
+                address
+            );
+        } else {
+            panic!("Multiple mutable borrows of an index at {:?}", address);
+        }
         view_changes
     }
 
-    /// Returns a mutable reference to the changes corresponding to a certain index.
+    /// Clones changes for a specific `View` from the patch. Panics if the changes
+    /// are mutably borrowed.
+    fn clone_view_changes(&self, address: &IndexAddress) -> Rc<ViewChanges> {
+        let mut changes = self.changes.borrow_mut();
+        changes
+            .entry(address.clone())
+            .or_insert_with(|| Some(Rc::new(ViewChanges::new())))
+            .as_ref()
+            .unwrap_or_else(|| {
+                panic!(
+                    "Attempting to borrow {:?} immutably while it's borrowed mutably",
```

Comment:
This comment looks weird. We've borrowed changes mutably several lines above, who exactly tries to borrow them immutable here?